### PR TITLE
New python bindings using pybind11

### PR DESCRIPTION
**Pull request by maxnoe**
Here I replace the existing python bindings which are not working with python3 with new bindings using the excellent pybind11 library.

TODO (for @HealthyPear): 

- [x] Add a python build system (`setup.py / pyproject.toml` etc for making this `pip install`able).
- [x] Add tests
- [x] Setup github actions? 

I for now removed the histogram class, which was pure python 2 and seemed unrelated to the xcdf IO functionality, let me know what you think.

Also if you think that anything in the API is missing.

### CODE REVIEW

File: .gitignore
Comment (maxnoe) [inline .gitignore:1]: Editor and IDE stuff should go into your personal, user-wide gitignore. Otherwise projects would have to add gitignores for each and every editor in use

File: CMakeLists.txt
Comment (maxnoe) [inline CMakeLists.txt:3]: It's usually much better to switch to the NEW behavior

File: CMakeLists.txt
Comment (maxnoe) [inline CMakeLists.txt:75]: Do not use both `add_subdirectory` and `find_package` for pybind11

File: src/pybindings/xcdf/Histogram.py
Comment (maxnoe) [reply src/pybindings/xcdf/Histogram.py:35]: We now have numpy in the requirements, so just import

File: CMakeLists.txt
Comment (HealthyPear) [reply CMakeLists.txt:3]: I decided to keep the old one because I didn't want to modify the current code at
https://github.com/jimbraun/XCDF/blob/176628a1373ae6c21d6cff9ed139676ed527a442/CMakeLists.txt#L39
which could be used in more places elsewhere

File: CMakeLists.txt
Comment (jimbraun) [inline CMakeLists.txt:45]: The new version should be 3.1.0, not 4.0.0.

Comment (maxnoe) [reply CMakeLists.txt:45]: The python interface has incompatible changes.

We could decouple the python interface Version from the cpp library Version.

File: .gitignore
Comment (HealthyPear) [reply .gitignore]: fixed

File: src/pybindings/xcdf/Histogram.py
Comment (HealthyPear) [reply src/pybindings/xcdf/Histogram.py]: fixed

File: CMakeLists.txt
Comment (maxnoe) [reply CMakeLists.txt:75]: by doing `find_package(pybind11 REQUIRED)` later, it already is.

Comment (maxnoe) [reply CMakeLists.txt:75]: https://stackoverflow.com/a/65176396/3838691

Says to prefer `find_package`.

Comment (maxnoe) [reply CMakeLists.txt:75]: Using both at the same time just doesn't make sense! If you use `add_subdirectory`,  the whole `IF(SKBUILD)` and having pybind11 in pyproject.toml doesn't make any sense.

Either use `find_package` and the whole skbuild shebang OR use `add_subdirectory`

Comment (HealthyPear) [reply CMakeLists.txt:75]: yes, I am suggesting to remove `find_package ` and keep `add_subdirectory `

**Comment by HealthyPear** [comment]
Hi @jimbraun,

I added a CMake-based continuous integration pipeline for the C++ integration with python bindings (working on the pure-python one).

You should be able to activate it and make it a requirement for merging from ``settings``-->``branches``-->``Branch protection rules`` (for the ``master``branch) --> ``Require status checks to pass before merging``--> ``Status checks``

File: CMakeLists.txt
Comment (maxnoe) [inline CMakeLists.txt:99]: not needed anymore after removing the old python bindings headers

File: .github/workflows/CI_cmake_build.yml
Comment (maxnoe) [inline .github/workflows/CI_cmake_build.yml:38]: There are tests, the CI should run them

Comment (HealthyPear) [reply .github/workflows/CI_cmake_build.yml:47]: done (for the CMake CI for the moment)

**Comment by HealthyPear** [comment]
Dear @jimbraun ,

I added also a CI based entirely on Python, so you can add also that to the required status checks for merging.
Unfortunately we have a problem in the macos installation (ubuntu works fine) so while the CMake-based pipeline will succeed the Python one will fail for the moment.

Next steps:

- [ ] fix macos installation
- [ ] add pytest(s), at least one in which we read a test file created with the C++ library
- [ ] check/add missing API

File: CMakeLists.txt
Comment (HealthyPear) [reply CMakeLists.txt:45]: The major version needs to be bumped up if the versioning refers to the whole package, as the Python API will cause a breaking change with respect to v3.00.03.

If decoupling is fine for @jimbraun, it's fine for me.
Personally, I think it would increase maintenance, and that going for v4.0.0 altogether is easier.

Comment (jimbraun) [reply CMakeLists.txt:45]: Yes, you are right, let's bump the major version to 4.0.0.

**Comment by HealthyPear** [comment]
Dear @jimbraun ,

I now added a (simple) Python testing function to make sure we correctly read an XCDF file written by the C++ interface.
The macOS installation issue should be fixed, but I could test it only manually.

It seems a good time to enable the CI pipelines (neither me or my colleague can, since we are not administrators):
`settings-->branches-->Branch protection rules (for master only is sufficient) --> Require status checks to pass before merging--> Status checks`.

The CI pipelines are 2:
- CMake build with Python bindings
- Python package

There are still few things missing like, e.g.
- a `with` statement to open a file in a context manager,
- a Close() method
- a method to restart the reading after the last event has been read
- exposing also the attributes of `XCDFFieldDescriptor` and not only a `__repr__` method might be useful
- +...?

**Comment by jimbraun** [comment]
@HealthyPear : Thanks much for adding the CI pipelines.  I agree and I'll enable these next week.

File: .github/workflows/CI_Python.yml
Comment (maxnoe) [inline .github/workflows/CI_Python.yml:25]: Run the pytests?

**Comment by jimbraun** [comment]
I merged the CI pipelines manually into master, and checks are now required to pass before merging.  Note that the Python checks currently fail in master, but merging this PR should be the next action in XCDF, so I feel this is OK for now.

File: .github/workflows/CI_Python.yml
Comment (HealthyPear) [reply .github/workflows/CI_Python.yml:25]: done

File: .github/workflows/CI_Python.yml
Comment (maxnoe) [inline .github/workflows/CI_Python.yml:26]: do `pip install ".[tests]"` above, then you can remove this line and it's future proof if you add more `tests` dependencies

File: .github/workflows/CI_cmake_build.yml
Comment (maxnoe) [inline .github/workflows/CI_cmake_build.yml:5]: This looks like its the same file as on master, just named `CI_cmake_build.yml` instead of `CI_make_build.yml`

File: .github/workflows/CI_Python.yml
Comment (HealthyPear) [reply .github/workflows/CI_Python.yml:26]: right - I forgot about the `setup.cfg`...

File: .github/workflows/CI_cmake_build.yml
Comment (HealthyPear) [reply .github/workflows/CI_cmake_build.yml:5]: ah ok now I understand from where the merge conflict arose...probably @jimbraun copy-pasted the workflows here in master

I think it would have been sufficient to activate them from the repo settings

Comment (HealthyPear) [reply .github/workflows/CI_cmake_build.yml:5]: and this probablt also explains why the cmake tests are all doubled

I can remove the one in master with a subsequent PR

File: .github/workflows/CI_Python.yml
Comment (HealthyPear) [reply .github/workflows/CI_Python.yml]: done

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [inline src/pybindings/xcdf/xcdf.cpp:25]: I'd give this a path as argument: `write_test_file(const std::string& path)` and then use the `tmp_path` fixture to not create it in the current directory.

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [inline src/pybindings/xcdf/xcdf.cpp:102]: RAII. This should not be necessary but be call `Close` here, it should be called on exit by the destructor.

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [reply src/pybindings/xcdf/xcdf.cpp:25]: It should also be `void`, there is no point to return 0 here, it is not a main function.

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [inline src/pybindings/xcdf/xcdf.cpp:174]: to make things nicer, you can provide docstrings and `py::arg` specifications for all the calls here.

File: setup.cfg
Comment (maxnoe) [inline setup.cfg:28]: numpy should be in `install_requires`.

File: src/pybindings/xcdf/xcdf.cpp
Comment (HealthyPear) [reply src/pybindings/xcdf/xcdf.cpp:102]: My C++ is quite rusty, but I don't see an automatic call to a destructor in the code of XCDF.cc...

Here when I call

`XCDFFile f("test_file_for_pybindings.xcd", "w");`

it should just mean `XCDFFile::Init`

`write_test_file` is basically the first half of `tests/SimpleTest.cc` and also there the Close method is explicitly called

File: setup.cfg
Comment (jimbraun) [reply setup.cfg:28]: I'm personally OK removing Histogram.py.  I'm reasonably certain that no one is currently using it, and, as you say, if anyone is using it, they should change to something more modern.

Comment (maxnoe) [reply setup.cfg:28]: @HealthyPear, no, the pybind11 code also returns data as numpy arrays, which will fail if it is not installed.

Comment (maxnoe) [reply setup.cfg:28]:
```
In [6]: f = xcdf.File('./test_file_for_pybindings.xcd', 'r')

In [7]: f
Out[7]: <xcdf.xcdf.File at 0x7fc8fe8a9830>

In [8]: next(f)
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In [8], line 1
----> 1 next(f)

ModuleNotFoundError: No module named 'numpy'
```

Comment (HealthyPear) [reply setup.cfg]: right. - fixed

File: src/pybindings/xcdf/xcdf.cpp
Comment (HealthyPear) [reply src/pybindings/xcdf/xcdf.cpp:170]: I didn't add docstrings because the pybind11 talks only about sphinx and I was actually planning to use doxygen (I never used sphinx outside of python)

regarding `py::arg` you mean [this](https://pybind11.readthedocs.io/en/stable/basics.html?highlight=py%3A%3Aarg#keyword-arguments)?

Comment (maxnoe) [reply src/pybindings/xcdf/xcdf.cpp:170]: Docstrings are needed! They are shown within python, e.g. when doing `help(<thing>)` or when looking at reprs, right now it just shows:

```
In [1]: from xcdf import File

In [2]: File
Out[2]: xcdf.xcdf.File

In [3]: File?
Docstring:      <no docstring>
Init docstring: __init__(self: xcdf.xcdf.File, path: str, mode: str = 'r') -> None
File:           ~/Projects/XCDF/src/pybindings/xcdf/xcdf.cpython-310-x86_64-linux-gnu.so
Type:           pybind11_type
Subclasses:    
```

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [inline src/pybindings/xcdf/xcdf.cpp:169]: See here for `py::arg` usage.

File: src/pybindings/xcdf/xcdf.cpp
Comment (HealthyPear) [reply src/pybindings/xcdf/xcdf.cpp:170]: > Docstrings are needed! 

of course :)

How do you write docstrings in an sowftware independend way in this .cpp file?
can I still do it as they show [here](https://pybind11.readthedocs.io/en/stable/advanced/misc.html?highlight=docstring#generating-documentation-using-sphinx) even though I am not sure if sphinx will be used in the end?

**Comment by HealthyPear** [comment]
@jimbraun I don't know why, but now it seems I need you to approve the workflows (they were running with not problem before...) - see [here](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks#approving-workflow-runs-on-a-pull-request-from-a-public-fork) (I hope it's a one-time operation)

File: src/pybindings/xcdf/xcdf.cpp
Comment (maxnoe) [reply src/pybindings/xcdf/xcdf.cpp:170]: You need to write them for python! So use the usual syntax. That is independent of whether they in the end are rendered using sphinx.

They are just given as argument to the pybind11 methods.

**Comment by maxnoe** [comment]
@HealthyPear I guess it's because I pushed changes and I wasn't approved yet

**Comment by jimbraun** [comment]
@HealthyPear : I've been manually approving all the workflow runs.  They don't start automatically for first-time contributors.

File: src/pybindings/xcdf/xcdf.cpp
Comment (HealthyPear) [reply src/pybindings/xcdf/xcdf.cpp]: done

File: src/pybindings/xcdf/xcdf.cpp
Comment (HealthyPear) [reply src/pybindings/xcdf/xcdf.cpp]: done

**Comment by HealthyPear** [comment]
@jimbraun could please trigger the workflows?
@maxnoe do you see something else that needs to be fixed in this PR?

File: setup.cfg
Comment (maxnoe) [inline setup.cfg:18]: 3.11 was released in the mean time

Comment (HealthyPear) [reply setup.cfg:18]: added

**Comment by HealthyPear** [comment]
@jimbraun could you do it again? I think we are done @maxnoe?

For the moment I think this PR has what needs to read XCDF data to Python3 in a basic way, aside from Python packages like astropy.

**Comment by HealthyPear** [comment]
@maxnoe if everything is ready I think you have to change the status of the PR from Draft

**Review by jimbraun — APPROVED** [review]
@HealthyPear and @maxnoe: I'm happy to merge if you feel this PR is ready.

**Comment by maxnoe** [comment]
From my side good!

**Comment by HealthyPear** [comment]
I am also satisfied!

**Comment by jimbraun** [comment]
Thanks @HealthyPear and @maxnoe!

**Pull request merged**
